pyvista/examples/_example_loader.py

```diff
@@ -0,0 +1,504 @@
+"""Abstraction layer for downloading, reading, and loading example files."""
+
+from __future__ import annotations
+
+from abc import abstractmethod
+import functools
+import os
+from typing import (
+    Any,
+    Callable,
+    Optional,
+    Protocol,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    runtime_checkable,
+)
+
+import pyvista
+from pyvista.core._typing_core import NumpyArray
+from pyvista.core.utilities.fileio import get_ext
+
+# The following classes define an API for working with either
+# a single filename or multiple filenames
+
+_T = TypeVar('_T')
+_FilePropStrType = TypeVar('_FilePropStrType', str, Tuple[str, ...], covariant=True)
+
+
+class _FileProps(Protocol[_FilePropStrType]):
+    @property
+    @abstractmethod
+    def filename(self) -> _FilePropStrType:
+        """Return the filename(s) of all files."""
+
+    @property
+    def extension(self) -> Union[str, Tuple[str, ...]]:
+        """Return the file extension(s) of all files."""
+        return _get_extension_from_filename(self.filename)
+
+    @property
+    @abstractmethod
+    def _filesize_bytes(self):
+        """Return the file size(s) of all files in bytes."""
+
+    @property
+    @abstractmethod
+    def _filesize_format(self) -> _FilePropStrType:
+        """Return the formatted size of all file(s)."""
+
+    @property
+    @abstractmethod
+    def total_size(self) -> str:
+        """Return the total size of all files."""
+
+    @property
+    @abstractmethod
+    def reader(self):
+        """Return the base file reader(s) used to read the files."""
+
+
+@runtime_checkable
+class _Downloadable(Protocol[_FilePropStrType]):
+    """...."""
+
+    @abstractmethod
+    def download(self) -> _FilePropStrType:
+        """...."""
+
+
+@runtime_checkable
+class _Loadable(Protocol):
+    """...."""
+
+    @abstractmethod
+    def load(self) -> Any:
+        """...."""
```

Comment:
@user27182
To understand the code, I added a docstring("""....""") to the section I want you to add. Please add there the intent of adding the class and method for our understanding.